enhancements/workload-partitioning/cpu-limits-workload-partitioning.md

```diff
@@ -0,0 +1,480 @@
+
+Currently this feature is not supported Hypershift or hosted control planes
+since NTO and the underlying mechanism need access to APIs that are not exposed
+via this type of topology.
+
+#### Standalone Clusters
+
+The overarching feature is already available on standalone clusters, this
+enhancement extends the existing implementation to support CPU limit, there
+should be no extra consideration for this type of topology outside of the
+described implementation.
+
+#### Single-node Deployments or MicroShift
+
+This enhancement is not adding any extra components to the existing feature, the
+original implementation was designed for Single Node as such there is no extra
+consideration outside of the described implementation.
+
+We are not adding any API changes as such MicroShift should not be effected by
+this change.
+
+### Risks and Mitigations
+
+The addition of this change should not pose any major problems, with this
+approach we are essentially emulating what Kubernetes does when limits are set.
+
+A thing to note is this is under the assumption that CPU Period is fixed at `100
+milliseconds`, from my investigations that does seem to be the case for RHCOS in
+OCP, but I am not sure if there are deployments out there that run with a
+different CPU period, those would need to be taken into account.
+
+Currently workload partitioning is only used by platform pods, any platform pods
+that use `limits` will correctly be modified to have limits and request be
+applied via `cpulimit` and `cpushares`. Since one of the key attributes of
+workload partitioning is pinning a workload to a specific CPU set, then those
+pods will correctly be moved over to those CPU sets and have limits imposed this
+might cause issues in performance for those pods.
+
+### Drawbacks
+
+A draw back is that we will be committing to carry the admission webhook patch
+downstream indefinitely. Futhermore, since we are not doing anything more than
+what the `limits` would do in Kubernetes, we are essentially re-applying that
+functionality and using our own existing workloads mechanism to pass the
+`limits` information down to the container runtime. This doubling of efforts is
+currently required in order to avoid the default behavior of any missing
+`limits.cpu` being applied to `requests.cpu`, since such an upstream change
+would not be favorable.
+
+## Design Details
+
+## Open Questions [optional]
+
+N/A
+
+## Test Plan
+
+We will update the origin tests for workload partitioning to also include checks
+for cpu limits on clusters where workload partitioning is enabled.
+
+## Graduation Criteria
+
+### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
```

Comment:
Do we have a OpenShift FeatureGate defined yet? We probably need one.

Reply:
Oh apologies @rphillips, that must have come along with the re-templating I did, I don't think those are relevant for this feature since this is extending the existing feature. Since the core feature is already GA all we are doing in this EP is allowing it correctly apply Limits. What do you think?